File: leetcode/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        """
        RUNTIME: 26 ms (97.89%)
        MEMORY: 13.8 MB (98.51%)
        """
        if len(s) % 2 == 1: return False
        pairs = {'(':')', '{':'}', '[':']' }
        stack = []
        for char in s:
            if char in pairs:
                stack.append(char)
            elif len(stack) == 0 or pairs[stack.pop()] != char: return False
        return len(stack) == 0

File: leetcode/0020-valid-parentheses/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 0 ms (100%)
     * MEMORY: 6.4 MB (52.7%)
    */
    bool isValid(string s) {
        if (s.length() % 2 == 1) {
            return false;
        }
        stack<char> st;
        for(char c : s){
            if(c == '('|| c == '{' || c == '[') {
                st.push(c);
            }
            else {
                if(st.empty()) return false;
                if(c == ')' && st.top() != '(') return false;
                if(c == '}' && st.top() != '{') return false;
                if(c == ']' && st.top() != '[') return false;
                st.pop();
            }
        }
        return st.empty();
    }
};
